### PR TITLE
[InstSimplify] Fix incorrect poison propagation when folding phi

### LINKED ISSUE #68683: [InstructionSimplify] InstructionSimplify error deleted `phi`
I tried this IR:

```llvm
define i32 @src(i1 %c0, i1 %c1, i32 %v1) {
start:
  br i1 %c0, label %bb0, label %bb1

bb0:
  br label %bb1

bb1:
  %x = phi i32 [ %v1, %bb0 ], [ undef, %start ]
  br i1 %c1, label %bb0, label %bb2

bb2:
  ret i32 %x
}
```

`phi` should not be deleted by InstructionSimplify.

alive2: https://alive2.llvm.org/ce/z/LNXeGp

### PR DESCRIPTION
**Pull request by nikic**
We can only replace phi(X, undef) with X, if X is known not to be poison. Otherwise, the result may be more poisonous on the undef branch.

Fixes https://github.com/llvm/llvm-project/issues/68683.

### CODE REVIEW

**Comment by dtcxzyw** [comment]
See https://github.com/llvm/llvm-project/issues/68683#issuecomment-1948117173.

**Comment by nikic** [comment]
@dtcxzyw Could you check this on llvm-opt-benchmark first? I'm wondering whether this introduces regressions that can be recovered based on more context.

The motivation for doing this *now* after ignoring this for so long is that we're definitely going to lose this when we stop using undef for uninit loads, so I'd like to know what the impact of just this part is going to be.

**Comment by dtcxzyw** [comment]
> @dtcxzyw Could you check this on llvm-opt-benchmark first? I'm wondering whether this introduces regressions that can be recovered based on more context.
> 
> The motivation for doing this _now_ after ignoring this for so long is that we're definitely going to lose this when we stop using undef for uninit loads, so I'd like to know what the impact of just this part is going to be.

Done.

**Comment by nikic** [comment]
@dtcxzyw Thanks! I think the main two takeaways are:

1. I suspect that we need a `nopoison` attribute (and I guess `!nopoison`) metadata. The current noundef is too strong, and in Rust often cannot be applied because e.g. parts of enums are only conditionally initialized. However, I believe that all frontend-generated arguments/returns/loads can be nopoison, because Rust does not expose poison in its semantics (and I think this holds for pretty much all other frontends as well). Having nopoison would mean that pretty much all "roots" can be known not to be poison. Once undef is removed (or at least not allowed to be referenced from inside functions) we would drop noundef and only have nopoison.
2. We should probably be folding `select c, undef, x` to `freeze(x)` if we can't determine that x is non-poison. Otherwise we may carry a complex select condition around all the way to the backend. Having the freeze is not ideal, but I think it's the better tradeoff. (The same kind of applies to phis after this change, but it's probably less important there because the condition is still used by the branch.)

**Comment by dtcxzyw** [comment]
> On a somewhat tangential note, I'm also wondering whether it would be worthwhile to add some flags on freeze so you can have freeze %x (freezes undef and poison -- used e.g. for despeculation), freeze undef %x (freezes undef only -- used for multi-use) and freeze poison %x (freezes poison only -- used e.g. for the select fold).

As we will finally get rid of undef, I don't think the (potential) benefit is worth the maintenance burden.

**Comment by nikic** [comment]
Moving this to draft as it looks like we're still not ready for this change :(

**Comment by nikic** [comment]
Unfortunately, we just found the first instance where this miscompile occurs in the wild (https://github.com/llvm/llvm-project/issues/97702).

Per the comments of @RalfJung in https://discourse.llvm.org/t/rfc-add-nopoison-attribute-metadata/79833 it looks like Rust does not want to use `nopoison` attributes, which basically leaves us no choice but to move forward with this as-is and have Rust eat the resulting codegen regressions.

**Comment by RalfJung** [comment]
For Rust, in the future we'd like to have poison semantics for our values, and then we cannot use `nopoison`. More precisely, today we could use `nopoison`, but it would be a shame if LLVM wouldn't let its frontends have poison semantics for values without codegen regressions. Alternatives exist, as mentioned in [my comment](https://discourse.llvm.org/t/rfc-add-nopoison-attribute-metadata/79833/12?u=ralfjung).

Today, Rust doesn't yet have values with poison semantics (it is regularly requested but we are blocked on lack of support in LLVM). So today, `nopoison` could be added in Rust. So if this is some sort of transition strategy that's perfectly fine, but it is not a great long-term solution.

**Review by dianqk — APPROVED** [review]
LGTM. Soundness is always the most important thing for me.

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (RalfJung) [reply llvm/lib/Analysis/InstructionSimplify.cpp:5286]: This is even a double-negation ("not guaranteed to not be poison"), so it's definitely not just you^^

**Comment by nikic** [comment]
Sorry, my initial analysis of https://github.com/llvm/llvm-project/issues/97702 wasn't right, and this wasn't the cause of that issue after all. So this one remains a theoretical problem only for now...

**Comment by dtcxzyw** [comment]
@nikic Can you provide some performance data on rust benchmarks before landing this patch?

**Comment by DianQK** [comment]
> @nikic Can you provide some performance data on rust benchmarks before landing this patch?

I'm adding this test.

**Comment by DianQK** [comment]
> @nikic Can you provide some performance data on rust benchmarks before landing this patch?

Based on the known test set, this has no noticeable impact: https://github.com/rust-lang/rust/pull/127347#issuecomment-2210100734.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Comment by nikic** [comment]
As far as I'm aware, this issue is still theoretical, but it does regularly cause confusion and duplicate issue reports, so I'm inclined to merge it.

**Review by dianqk — APPROVED** [review]

**Pull request merged**
